web/modules/custom/contact_form/src/EventSubscriber/NotificationSubscriber.php

```diff
@@ -76,19 +63,160 @@ public function contactForm(NotificationEvent $event) {
   }
 
   /**
-   * Returns the users with a given role.
+   * Determines the entity object of the route matching a uri.
    *
-   * @param string $role_id
-   *   The role id.
+   * @param string $uri
+   *   The internal or external uri.
+   *
+   * @return \Drupal\Core\Entity\EntityInterface|null
+   *   The entity object or null in any other case.
+   */
+  protected function getEntityFromUrl($uri) {
+    try {
+      $url = Url::fromUri($uri);
+    }
+    catch (Exception $e) {
+      return NULL;
+    }
+
+    if (!$url->isRouted()) {
+      return NULL;
+    }
+
+    $route_parameters = $url->getRouteParameters();
+    if (!isset($route_parameters['node']) && !isset($route_parameters['rdf_entity'])) {
+      return NULL;
+    }
+
+    // Check what type of content we are reporting.
+    $entity_type_id = NULL;
+    if (isset($route_parameters['node'])) {
+      $entity_type_id = 'node';
+    }
+    elseif (isset($route_parameters['rdf_entity'])) {
+      $entity_type_id = 'rdf_entity';
+    }
+    if ($entity_type_id === NULL) {
+      return NULL;
+    }
+
+    if ($url->getRouteName() !== "entity.{$entity_type_id}.canonical") {
+      return NULL;
+    }
+
+    $entity_id = $route_parameters[$entity_type_id];
+    $entity = $this->entityTypeManager->getStorage($entity_type_id)->load($entity_id);
+
+    return $entity;
+  }
+
+  /**
+   * Sends a report message instead of the normal contact form email.
+   *
+   * @param \Drupal\joinup_notification\Event\NotificationEvent $event
+   *   The event object.
+   */
+  protected function sendReportMessage(NotificationEvent $event) {
+    $message = $event->getEntity();
+    $entity = $this->getEntityFromUrl($message->get('field_contact_url')->first()->uri);
+    // If there is no entity found, do not send an email.
+    if (empty($entity)) {
+      return;
+    }
+    $user_data = $this->getReportUserData();
+    $user_data = $this->getUsersMessages($user_data, $entity);
+    $arguments = $this->generateReportArguments($message);
+    $this->sendUserDataMessages($user_data, $arguments);
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * Skip generating the arguments during the sending process.
+   */
+  protected function sendUserDataMessages(array $user_data, array $arguments = []) {
+    foreach ($user_data as $template_id => $user_ids) {
+      $values = ['template' => $template_id, 'arguments' => $arguments];
+      $message = Message::create($values);
+      $message->save();
+
+      foreach ($user_ids as $user_id) {
```

Comment:
Never load entity by entity for performance reasons. We should use here: `foreach ($this->entityTypeManager->getStorage('user')->loadMultiple($user_ids) as $user_id => $user)`.